File: backend/app/schemas/rank.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

class RankResponse(BaseModel):
    id: int
    name: str
    level: int
    team_turnover_required: float
    first_leg_requirement: float
    second_leg_requirement: float
    other_legs_requirement: float
    bonus_amount: float
    infinity_bonus_percentage: Optional[float]
    display_name: Optional[str]
    icon_url: Optional[str]
    color_hex: Optional[str]
    description: Optional[str]
    
    class Config:
        from_attributes = True

class RankProgress(BaseModel):
    current_rank: RankResponse
    next_rank: Optional[RankResponse]
    total_turnover: float
    total_turnover_progress: float
    first_leg_turnover: float
    first_leg_progress: float
    second_leg_turnover: float
    second_leg_progress: float
    other_legs_turnover: float
    other_legs_progress: float
    overall_progress: float
    requirements_met: dict

class RankHistory(BaseModel):
    rank_name: str
    achieved_date: datetime
    bonus_earned: float

File: frontend/src/types/rank.ts
export interface RankResponse {
  id: number;
  name: string;
  level: number;
  team_turnover_required: number;
  first_leg_requirement: number;
  second_leg_requirement: number;
  other_legs_requirement: number;
  bonus_amount: number;
  infinity_bonus_percentage?: number;
  display_name?: string;
  icon_url?: string;
  color_hex?: string;
  description?: string;
}

export interface RankProgress {
  current_rank: RankResponse;
  next_rank?: RankResponse;
  total_turnover: number;
  total_turnover_progress: number;
  first_leg_turnover: number;
  first_leg_progress: number;
  second_leg_turnover: number;
  second_leg_progress: number;
  other_legs_turnover: number;
  other_legs_progress: number;
  overall_progress: number;
  requirements_met: Record<string, boolean>;
}

export interface RankHistory {
  rank_name: string;
  achieved_date: string;
  bonus_earned: number;
}
